scylla/src/transport/query_result.rs

```diff
@@ -0,0 +1,210 @@
+use bytes::Bytes;
+use thiserror::Error;
+use uuid::Uuid;
+
+use scylla_cql::frame::frame_errors::ParseError;
+use scylla_cql::frame::response::result::{ColumnSpec, RawRows, Row};
+use scylla_cql::types::deserialize::row::{DeserializeRow, TypedRowIterator};
+
+use super::legacy_query_result::Legacy08QueryResult;
+
+/// Raw results of a single query.
+///
+/// More comprehensive description TODO
+#[derive(Default, Debug)]
+pub struct QueryResult {
+    raw_rows: Option<RawRows>,
+    tracing_id: Option<Uuid>,
+    warnings: Vec<String>,
+}
+
+impl QueryResult {
+    /// Returns the number of received rows, or `None` if the response wasn't of Rows type.
+    pub fn rows_num(&self) -> Option<usize> {
+        Some(self.raw_rows.as_ref()?.rows_count())
+    }
+
+    /// Returns a bool indicating the current response is of Rows type.
+    pub fn is_rows(&self) -> bool {
+        self.raw_rows.is_some()
+    }
+
+    /// Returns the received rows when present.
+    ///
+    /// Returns an error if the original query didn't return a Rows response (e.g. it was an `INSERT`),
+    /// or the response is of incorrect type.
+    pub fn rows<'s, R: DeserializeRow<'s>>(&'s self) -> Result<TypedRowIterator<'s, R>, RowsError> {
+        Ok(self
+            .raw_rows
+            .as_ref()
+            .ok_or(RowsError::NotRowsResponse)?
+            .rows_iter()?)
+    }
+
+    /// Returns the received rows when present, or None.
+    ///
+    /// Returns an error if the rows are present but are of incorrect type.
+    pub fn maybe_rows<'s, R: DeserializeRow<'s>>(
+        &'s self,
+    ) -> Result<Option<TypedRowIterator<'s, R>>, ParseError> {
+        match &self.raw_rows {
+            Some(rows) => Ok(Some(rows.rows_iter()?)),
+            None => Ok(None),
+        }
```

Comment:
I tend to agree with @cvybhu here. Although I like the functional style myself, combinators like `and_then`, `map`, `transpose` are good only in moderation. Understanding an expression that uses those combinators heavily requires shuffling a lot of context in your head - you need to manually infer the type of the expression after each dot in order to properly understand what it does. This is especially tricky here because of the nesting of `Option` and `Result` layers. Writing such expressions is a lot of fun, and reading them is a nice mental exercise, but I think it's only detrimental in production code.

Actually, the `QueryResult::maybe_first_row` that I wrote also suffers from the same problem - I'll try to make it more understandable.